src/NHapi.Base/Util/SynchronizedCache.cs

```diff
@@ -0,0 +1,144 @@
+﻿#if NETFRAMEWORK
+
+namespace NHapi.Base.Util
+{
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Threading;
+
+    internal class SynchronizedCache<TKey, TValue> : IEnumerable<KeyValuePair<TKey, TValue>>
+    {
+        private readonly ReaderWriterLockSlim readerWriterLock = new ReaderWriterLockSlim();
+        private readonly Dictionary<TKey, TValue> dictionary = new Dictionary<TKey, TValue>();
+
+        ~SynchronizedCache()
+        {
+            if (readerWriterLock != null)
+            {
+                readerWriterLock.Dispose();
+            }
+        }
+
+        public int Count => dictionary.Count;
```

Comment:
Does this need to be protected by a read lock?

Reply:
Possibly, I'll add it anyway.

Reply:
Done!